Rewrite with ScreenCaptureKit

karaggeorge opened the pull request:
Rewrites Aperture to use ScreenCaptureKit

Breaking changes:
- Now requires macOS 13+
- `highlightClicks` is only available on macOS 15+
- Aperture itself is not the recorder, you can instantiate `Aperture.Recorder()` and use `startRecording()` and `stopRecording()` on it
- Everything uses async/await instead of callbacks
- `startRecording` awaits until the file actually starts getting written before resolving, or throws any errors that happen during that time
- `cropRect` is top/left now (this is how `ScreenCaptureKit` works by default)

This has as much feature parity as possible (other then the `highlightClicks` note above) but it adds some functionality as well:
- Also supports `.mov` and `.m4v` for videos
- Recording system audio (+ mic)
- Loseless audio option
- Recording audio only (only `.m4a` for now, AAC or ALAC encoding)
- Recording a specific window (will follow the window around even if it moves or covered by other windows)

Potential features we'd want to add in a follow-up:
- Support more audio formats
- Maybe support quality/bitrate options for audio/video
- Support recording applications as targets
- Support excluding apps/windows from the recording (for screen/application only)
	- Some good default sets would be excluding the current application itself and/or desktop icons

This PR still needs to update the readme, but wanted to get it reviewed first to fully land on the implementation before I go and update the readme

Closes #16 
Closes #31
Closes #71 
Closes #63 
Closes #65


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#65: Support for top/left positions for capture](https://oss.issuehunt.io/repos/65488992/issues/65)
---
</details>
<!-- /Issuehunt content-->

karaggeorge committed:
Rewrite with ScreenCaptureKit

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
+				guard let window = content.windows.first(where: { window in
+					String(window.windowID) == options.targetId
+				}) else {
+					throw ApertureError.targetNotFound(options.targetId!)
```
Don't force-unwrap

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
+				filter = screenFilter
+				break
+			case .window:
+				/// We need to cal this before `SCContentFilter` below otherwise an error is thrown: https://forums.developer.apple.com/forums/thread/743615
```
```suggestion
				/// We need to call this before `SCContentFilter` below otherwise an error is thrown: https://forums.developer.apple.com/forums/thread/743615
```

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
+				break
+			}
+			
+			if let filter = filter {
```
```suggestion
			if let filter {
```

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
+		public func stopRecording() async throws {
+			if let error = error {
+				throw error
 			} else {
```
The `else` is moot here.

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
 			} else {
-				throw Error.couldNotAddMic
+				onFinish?()
+				try? await cleanUp()
```
Should it not clean up on error?

karaggeorge replied:
The only way we have an error here is if an error was thrown after starting, and `recordError` does the cleanup (in case stop is never called)

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
+			systemAudioInput?.markAsFinished()
+			microphoneInput?.markAsFinished()
+			
+			if let microphoneCaptureSession = microphoneCaptureSession, microphoneCaptureSession.isRunning {
```
```suggestion
			if microphoneCaptureSession?.isRunning == true {
```

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
+				externalDeviceCaptureSession.stopRunning()
+			}
+			
+			if (assetWriter?.status == .writing) {
```
```suggestion
			if assetWriter?.status == .writing {
```

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
-			audioDevice: audioDevice,
-			videoCodec: videoCodec
-		)
+extension Aperture.Recorder: SCStreamDelegate, SCStreamOutput, AVCaptureAudioDataOutputSampleBufferDelegate, AVCaptureVideoDataOutputSampleBufferDelegate {
```
If possible, one extension per conformance. Makes it easier to read.

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
+	private func initOutput(target: Aperture.Target, options: Aperture.RecordingOptions, streamConfig: SCStreamConfiguration) async throws {
+		let assetWriter = try getAssetWriter(target: target, options: options)
+		
+		var audioSettings: [String: Any] = [AVSampleRateKey : 48000, AVNumberOfChannelsKey : 2]
```
```suggestion
		var audioSettings: [String: Any] = [AVSampleRateKey: 48000, AVNumberOfChannelsKey: 2]
```

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
+		case unknownError(Swift.Error)
+		case noPermissions
+	}
+	public final class Recorder: NSObject {
```
Define this in an extension for improved readability.

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
+		
+		/// Internal helpers for when we are resuming, used to fix the buffer timing
+		private var isResuming: Bool = false
+		private var timeOffset: CMTime = .zero
```
```suggestion
		private var timeOffset = CMTime.zero
```

karaggeorge committed:
Fix scaleFactor

sindresorhus commented on Sources/Aperture/Utilities.swift:
```diff
@@ -1,98 +1,65 @@
-
-			service = IOIteratorNext(iterator)
+extension Aperture {
+	public static func hasPermissions() async -> Bool {
```
```suggestion
	public static var hasPermissions: Bool {
		get async {
```

sindresorhus commented on Sources/Aperture/Utilities.swift:
```diff
@@ -1,98 +1,65 @@
-			return "Unnamed screen"
+extension CMSampleBuffer {
+	public func adjustTime(by offset: CMTime) -> CMSampleBuffer? {
+		guard CMSampleBufferGetFormatDescription(self) != nil else { return nil }
```
Use `self.formatDescription`. Same with the other things.

karaggeorge commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
+	public struct RecordingOptions {
+		public init(
+			destination: URL,
+			targetId: String? = nil,
```
Thanks! These are great. Took some courses on Swift before I wrote this to get a better understanding, but they don't really cover formatting

Is there something like eslint/prettier for Swift that can help me identify these? (and things like not adding `: Bool` if there's a default value etc)

sindresorhus replied:
`swiftlint` with this config: https://github.com/sindresorhus/swiftlint-config

karaggeorge commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,662 @@
+		private var activity: NSObjectProtocol?
+		
+		/// Whether the recorder has started writting to the output file
+		private var isRunning: Bool = false
+		/// Whether the recorder is paused
+		private var isPaused: Bool = false
```
How can I do this with an `enum`? Do you mean `struct` to hold each var? They can have independent values at times

sindresorhus replied:
Can it be `isRunning == true` and `isPaused == true` at the same time?

karaggeorge committed:
Address PR feedback

karaggeorge committed:
Add swiftlint config

karaggeorge committed:
swiftlint

karaggeorge commented:
@sindresorhus addressed all except for [this](https://github.com/wulkano/Aperture/pull/80#discussion_r1835767003) and configured/ran swiftlint 😌

karaggeorge committed:
Call onStart

karaggeorge committed:
Make isPaused public get

karaggeorge committed:
Fix onFinish order

karaggeorge commented:
While working on `aperture-node` with this version, I noticed a few things missing. Pushed some commits above, but need to also add:

- Support for proRes codecs 
- Error handling if we can't add the video inputs (we silently fail right now, took me a while to figure out why prores wasn't working)

will be pushing a commit with those two at some point tomorrow

karaggeorge committed:
Add support for proRes, improve error handling and cleanup

karaggeorge commented:
Ok, I think it's ready:
- Added support for proRes
	- Created a scoped enum for the ones we support and an extension to map them
	- Added checks for invalid file extensions <-> codec
- Added better error handling
	- Fixed some errors to have clearer message
	- Fixed logging for error sub-reasons
	- Throw errors if the stream does not start instead of failing silently or crashing
- Added a reset to the class so it can be re-used after recording is done

karaggeorge committed:
Fix example log

sindresorhus commented on Sources/Aperture/Devices.swift:
```diff
@@ -1,66 +1,141 @@
+		if let mode = CGDisplayCopyDisplayMode(self.displayID) {
+			return mode.pixelWidth / mode.width
+		}
+			return 1
```
```suggestion

		return 1
```

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
 
-	public func fileOutputShouldProvideSampleAccurateRecordingStart(_ output: AVCaptureFileOutput) -> Bool { true }
+		if output is AVCaptureVideoDataOutput {
+			if assetWriter != nil, !isRunning {
```
```suggestion
			if
				assetWriter != nil,
				!isRunning
			{
```

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
+		let sampleBuffer = handleBuffer(buffer: sampleBuffer, isVideo: output is AVCaptureVideoDataOutput)
+
+		if output is AVCaptureAudioDataOutput {
+			if assetWriter != nil && !isRunning && target == .audioOnly {
```
```suggestion
			if
				assetWriter != nil,
				!isRunning,
				target == .audioOnly
			{
```

Same in many places. Prefer `,` over `&&` in if/guard statements.

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
+
+extension Aperture.Recorder: AVCaptureAudioDataOutputSampleBufferDelegate, AVCaptureVideoDataOutputSampleBufferDelegate {
+	public func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
+		if isPaused {
```
```suggestion
		guard !isPaused else {
```

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
+
+extension Aperture.Recorder: SCStreamOutput {
+	public func stream(_ stream: SCStream, didOutputSampleBuffer sampleBuffer: CMSampleBuffer, of type: SCStreamOutputType) {
+		if isPaused {
+			return
+		}
+		guard sampleBuffer.isValid else {
+			return
+		}
```
Could be combined.

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
-		onResume?()
+	private func handleBuffer(buffer: CMSampleBuffer, isVideo: Bool) -> CMSampleBuffer {
+		/// Use the video buffer to syncronize after pausing if we have it, otherwise the audio
+		if !isVideo && target != .audioOnly {
```
Use `guard`.

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
-		shouldPreventSleep = false
-		onPause?()
+	private func startVideoStream(sampleBuffer: CMSampleBuffer) {
+		if target == .audioOnly {
```
`guard`

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
-			let connection = try output.connection(with: .video)
-				.unwrapOrThrow(Error.couldNotSetVideoCodec)
-			output.setOutputSettings([AVVideoCodecKey: videoCodec], for: connection)
+		private func reset() {
+			self.target = nil
+			self.options = nil
+
+			self.error = nil
+			self.continuation = nil
+			self.activity = nil
+
+			self.stream = nil
+			self.isStreamRecording = false
+
+			self.assetWriter = nil
+			self.videoInput = nil
+			self.systemAudioInput = nil
+			self.microphoneInput = nil
+
+			self.microphoneDataOutput = nil
+			self.externalDeviceDataOutput = nil
+
+			self.microphoneCaptureSession = nil
+			self.externalDeviceCaptureSession = nil
+
+			self.isRunning = false
+			self.isPaused = false
+			self.isResuming = false
+			self.timeOffset = CMTime.zero
+			self.lastFrame = nil
 		}
```
Having to reset this much state is kinda fragile. It would be easy to introduce more state and forget to reset it here. This is usually a sign that it needs some better encapulation. For example, a wrapper class that initiates this class on start record and destroys it on stop record. Then there is no need to reset.

karaggeorge replied:
Renamed this to `RecordingSession` and wrapped in a new `Recorder` class

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
-		get { activity != nil }
-		set {
-			if newValue {
-				activity = ProcessInfo.processInfo.beginActivity(options: .idleSystemSleepDisabled, reason: "Recording screen")
```
The activity handling could be simplified with something like this:

```swift
final class Activity {
	private let activity: NSObjectProtocol

	init(
		_ options: ProcessInfo.ActivityOptions = [],
		reason: String
	) {
		self.activity = ProcessInfo.processInfo.beginActivity(options: options, reason: reason)
	}

	deinit {
		ProcessInfo.processInfo.endActivity(activity)
	}
}
```

sindresorhus replied:
The benefit of the class is that it's self-contained you can just `nil` the property in the recorder to end the activity.

karaggeorge replied:
Sounds good. Updated.

I think all the comments are addressed, I'm going to start working on the readme updates

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
+			}
+			if lastFrame.flags.contains(.valid) {
+				if timeOffset.value > 0 {
+					pts = CMTimeSubtract(pts, timeOffset)
```
CMTime support `+` and `-`

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
+		}
+
+		var lastFrame = CMSampleBufferGetPresentationTimeStamp(resultBuffer)
+		let dur = CMSampleBufferGetDuration(resultBuffer)
```
Don't use abbreviations

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
+		if isResuming {
+			isResuming = false
+
+			var pts = CMSampleBufferGetPresentationTimeStamp(buffer)
```
No abbreviations.

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,823 @@
+			resultBuffer = resultBuffer.adjustTime(by: timeOffset) ?? resultBuffer
+		}
+
+		var lastFrame = CMSampleBufferGetPresentationTimeStamp(resultBuffer)
```
Use `sampleBuffer` instance methods, not the legacy global methods. Applies in many places.

karaggeorge committed:
Fix formatting

karaggeorge committed:
lint

karaggeorge committed:
Expose isPaused through the recorder

karaggeorge committed:
Abstract activity

sindresorhus commented:
You don't need to `nil` it in the clean up. It will end when the `RecordingSession` class is `nil`'d. However, should the activity be ended when paused?

karaggeorge committed:
Update readme

karaggeorge committed:
Remove explicit activity nil

karaggeorge commented:
@sindresorhus  Updated 👍 

I think it's ok to keep during pause, not sure what would happen if it went to sleep, then resumed after

Also updated the readme. We didn't have docs before, added some basic usage and options

sindresorhus commented on readme.md:
```diff
@@ -1,28 +1,266 @@
-	<p align="center">Record the screen on macOS</p>
+ <img src="Media/aperture-logo.svg" width="64" height="64">
+ <h3 align="center">Aperture</h3>
+ <p align="center">Record the screen on macOS</p>
```
Incorrect indentation.

sindresorhus commented on readme.md:
```diff
@@ -1,28 +1,266 @@
+ options: Aperture.RecordingOptions(
+  destination: URL(fileURLWithPath: "./screen-recording.mp4"),
+  targetID: screen.id,
+ )
```
Incorrect indentation.

sindresorhus commented on readme.md:
```diff
@@ -1,28 +1,266 @@
+
+#### `highlightClicks`
+
+Type: `Bool`
+
+Default: `false`
```
```suggestion
Type: `Bool`\
Default: `false`
```

sindresorhus commented:
Instead of documenting the API in the readme, it would be better to use DocC and add good doc comments to the methods.

You just need https://github.com/sindresorhus/Defaults/blob/main/.spi.yml and get it added to https://swiftpackageindex.com/add-a-package

karaggeorge committed:
Fix error handling for errors during stream start

sindresorhus commented on Example/Sources/Example/main.swift:
```diff
@@ -7,32 +7,47 @@ func delay(seconds: TimeInterval, closure: @escaping () -> Void) {
 }
 
```
`delay` is unused.

sindresorhus commented on Example/Sources/Example/main.swift:
```diff
@@ -7,32 +7,47 @@ func delay(seconds: TimeInterval, closure: @escaping () -> Void) {
 }
 
 let url = URL(fileURLWithPath: "../screen-recording.mp4")
```
```suggestion
let url = URL(filePath: "../screen-recording.mp4")
```

sindresorhus commented on Example/Sources/Example/main.swift:
```diff
@@ -7,32 +7,47 @@ func delay(seconds: TimeInterval, closure: @escaping () -> Void) {
+	print("Finished recording:", url.path)
+	exit(0)
+} catch let error as Aperture.Error {
+	print("Aperture Error: \(error.localizedDescription)")
```
It should exit here too.

sindresorhus commented on Package.swift:
```diff
@@ -1,10 +1,10 @@
-// swift-tools-version:5.5
+// swift-tools-version:5.7
```
Why such an old Xcode version?

karaggeorge replied:
Didn't know that was an xcode version to be honest. I updated the minimum version to v13 and it said this has to be at least 5.7, so I changed it to fix

I'll update to 6.0.2

sindresorhus replied:
Make it "5.11". 6 will trigger a lot of concurrency things we don't want to deal with yet.

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,846 @@
+			do {
+				try await initOutput(target: target, options: options, streamConfig: streamConfig)
+			} catch {
+				let finalError: Aperture.Error
+				if let error = error as? Aperture.Error {
+					finalError = error
+				} else {
+					finalError = Error.couldNotStartStream(error)
+				}
```
```suggestion
				let finalError = if let error = error as? Error {
					error
				} else {
					.couldNotStartStream(error)
				}
```

sindresorhus commented on Sources/Aperture/Devices.swift:
```diff
@@ -1,66 +1,141 @@
+
+extension SCDisplay {
+	var nsScreen: NSScreen? {
+		NSScreen.screens.first { $0.displayID == self.displayID }
```
```suggestion
		NSScreen.screens.first { $0.displayID == displayID }
```

sindresorhus commented on Sources/Aperture/Devices.swift:
```diff
@@ -1,66 +1,141 @@
+	}
+
+	var scaleFactor: Int {
+		if let mode = CGDisplayCopyDisplayMode(self.displayID) {
+			return mode.pixelWidth / mode.width
+		}
+		return 1
```
```suggestion
		if let mode = CGDisplayCopyDisplayMode(displayID) {
			return mode.pixelWidth / mode.width
		}

		return 1
```

sindresorhus commented on Sources/Aperture/Devices.swift:
```diff
@@ -1,66 +1,141 @@
+		public struct Window: Hashable, Codable, Identifiable {
+			public let id: String
+			public let title: String?
+			public let applicationName: String?
+			public let applicationBundleIdentifier: String?
```
```suggestion
			public let appName: String?
			public let appBundleIdentifier: String?
```

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,846 @@
+		/// The error handler for the recording session
+		var onError: ((Error) -> Void)?
+
+		func startRecording(
```
```suggestion
		func start(
```

It's inconsistent with `pause` and the recorder part is already clear from the namespace. `recorder.start()`

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,846 @@
+
+				filter = screenFilter
+			case .window:
+				/// We need to call this before `SCContentFilter` below otherwise an error is thrown: https://forums.developer.apple.com/forums/thread/743615
```
Only documentation comments should use `///`. Applies in many places.

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,846 @@
-			throw Error.couldNotAddScreen
+extension Aperture {
+	internal final class RecordingSession: NSObject {
+		/// The stream object for capturing anything on displays
```
```suggestion
		// The stream object for capturing anything on displays
```

sindresorhus commented on Sources/Aperture/Aperture.swift:
```diff
@@ -1,228 +1,846 @@
-		} else {
-			throw Error.couldNotAddScreen
+extension Aperture {
+	internal final class RecordingSession: NSObject {
```
```suggestion
	final class RecordingSession: NSObject {
```

`internal` is the default.

karaggeorge committed:
Address PR feedback

karaggeorge committed:
Revert spacing

karaggeorge committed:
Fix build

karaggeorge commented:
@sindresorhus re: docs, do I just need the `///` comments above the exported methods and to add that file? Any way to check what it looks like locally? 

I've added those for most properties, but that class is `internal` now. The "exposed" one only has start/stop/pause/resume and the `onStart` etc handlers

If it's a lot of work, can we do it in a follow-up?

sindresorhus commented on readme.md:
```diff
@@ -6,23 +6,254 @@
+
+guard let screen = screens.first else {
+  // No screens
+  exit(1)
```
Tab indentation, to match the code.

sindresorhus commented on readme.md:
```diff
@@ -6,23 +6,254 @@
+try await recorder.startRecording(
+  target: .screen,
+  options: Aperture.RecordingOptions(
+    destination: URL(fileURLWithPath: "./screen-recording.mp4"),
```
```suggestion
    destination: URL(filePath: "screen-recording.mp4"),
```

sindresorhus commented on readme.md:
```diff
@@ -6,23 +6,254 @@
+  exit(1)
+}
+
+try await recorder.startRecording(
```
```suggestion
try await recorder.start(
```

sindresorhus commented on readme.md:
```diff
@@ -6,23 +6,254 @@
+
+try await Task.sleep(for: .seconds(5))
+
+try await recorder.stopRecording()
```
```suggestion
try await recorder.stop()
```

sindresorhus commented on readme.md:
````diff
@@ -6,23 +6,254 @@
+```swift
+let screens = try await Aperture.Devices.screen()
+```
+<!-- Aperture.Devices.window(excludeDesktopWindows: false, onScreenWindowsOnly: false) -->
````
Leftover

karaggeorge committed:
Add docs and fix readme

karaggeorge commented:
@sindresorhus added in-line docs and added a link to the readme for swift package index (I kept the other readme sections as well)

I'm not sure if I can submit it to package index until after it's merged to main, but also figured the text might need changes

sindresorhus reviewed (approved):
Look good to me. Nice work! 👍

karaggeorge commented:
@sindresorhus Thanks for all the reviews! 

Is there anything special about releasing? Or just cut a release tag using the GH UI?
It looks like SPM clones these from GH, but I'm not sure
